File: api/rest/src/regulator.ts
/**
 * Regulator Portal Backend API
 *
 * Provides REST endpoints for:
 * - Querying audit trails with selective disclosure
 * - Managing data sharing agreements
 * - Verifying tamper-evidence chains
 * - Generating compliance reports
 * - Accessing regulatory audit logs
 */

import express, { Request, Response, NextFunction } from "express";
import { v4 as uuidv4 } from "uuid";

/**
 * Regulator authentication context
 */
interface RegulatorContext {
  regulatorId: string;
  role: "auditor" | "officer" | "admin";
  dsaId?: string;
  standards: Array<"ISA3000" | "SOC2" | "GDPR" | "SOX">;
}

/**
 * Audit trail query parameters
 */
interface AuditTrailQuery {
  startTime?: number;
  endTime?: number;
  eventTypes?: string[];
  submitter?: string;
  minSensitivity?: "public" | "internal" | "confidential" | "restricted";
  onlyControlEvents?: boolean;
  limit?: number;
  offset?: number;
}

/**
 * Selective disclosure request
 */
interface DisclosureRequest {
  eventIndex: number;
  allowedFields: string[];
  regulatorId: string;
}

/**
 * Data Sharing Agreement request
 */
interface DSARequest {
  dataProvider: string;
  regulatorAddress: string;
  standards: string[];
  allowedEventTypes: string[];
  role: string;
  effectiveLedger: number;
  expiryLedger?: number;
}

/**
 * Compliance report request
 */
interface ComplianceReportRequest {
  standard: "ISA3000" | "SOC2" | "GDPR" | "SOX";
  auditSubject: string;
  eventsExamined: number;
  objectivesTested: string[];
  controlsOperating: number;
  controlsDeficient: number;
}

/**
 * Regulator portal routes
 */
export function createRegulatorRoutes(): express.Router {
  const router = express.Router();

  /**
   * GET /regulator/audit-trails
   * Query immutable audit trail with optional selective disclosure
   */
  router.get("/regulator/audit-trails", async (req: Request, res: Response, next: NextFunction) => {
    try {
      const context = req.regulatorContext as RegulatorContext;
      if (!context) {
        return res.status(401).json({ error: "Unauthorized" });
      }

      const query: AuditTrailQuery = {
        startTime: req.query.startTime ? Number(req.query.startTime) : undefined,
        endTime: req.query.endTime ? Number(req.query.endTime) : undefined,
        eventTypes: req.query.eventTypes
          ? Array.isArray(req.query.eventTypes)
            ? req.query.eventTypes as string[]
            : [req.query.eventTypes as string]
          : undefined,
        submitter: req.query.submitter as string | undefined,
        minSensitivity: (req.query.minSensitivity as any) || "public",
        onlyControlEvents: req.query.onlyControlEvents === "true",
        limit: Math.min(Number(req.query.limit) || 100, 1000),
        offset: Number(req.query.offset) || 0,
      };

      // In production, this would query the smart contract via RPC
      const auditTrail = {
        query,
        regulatorId: context.regulatorId,
        standards: context.standards,
        entries: [
          {
            eventIndex: 0,
            eventHash: "0x" + "a".repeat(64),
            timestamp: Date.now(),
            eventType: "access_control",
            submitter: "GA...",
            sensitivity: "confidential",
            controlEvent: true,
          },
        ],
        totalCount: 1,
        hasMore: false,
      };

      res.json(auditTrail);
    } catch (error) {
      next(error);
    }
  });

  /**
   * GET /regulator/audit-trails/:eventIndex
   * Get a specific audit trail entry with optional disclosure proof
   */
  router.get("/regulator/audit-trails/:eventIndex", async (req: Request, res: Response, next: NextFunction) => {
    try {
      const context = req.regulatorContext as RegulatorContext;
      if (!context) {
        return res.status(401).json({ error: "Unauthorized" });
      }

      const eventIndex = Number(req.params.eventIndex);
      const includeProof = req.query.proof === "true";

      const entry = {
        eventIndex,
        eventHash: "0x" + "a".repeat(64),
        prevHash: "0x" + "b".repeat(64),
        timestamp: Date.now(),
        eventType: "access_control",
        category: "security",
        submitter: "GA...",
        metadata: includeProof ? undefined : { /* encrypted */ },
        version: 1,
        parentEventId: null,
        regulatory: {
          standard: "ISA3000",
          controlCode: "CC6.1",
          demonstratesControl: true,
          retentionLedgers: 52560,
          sensitivity: "confidential",
        },
      };

      if (includeProof) {
        entry.disclosure = {
          eventIndex,
          disclosedRoot: "0x" + "c".repeat(64),
          completeRoot: "0x" + "d".repeat(64),
          disclosedFields: ["timestamp", "eventType", "submitter"],
          merkleProof: [],
        };
      }

      res.json(entry);
    } catch (error) {
      next(error);
    }
  });

  /**
   * POST /regulator/selective-disclosure
   * Generate selective disclosure proof for an event
   */
  router.post("/regulator/selective-disclosure", async (req: Request, res: Response, next: NextFunction) => {
    try {
      const context = req.regulatorContext as RegulatorContext;
      if (!context) {
        return res.status(401).json({ error: "Unauthorized" });
      }

      const disclosureReq = req.body as DisclosureRequest;

      const proof = {
        eventIndex: disclosureReq.eventIndex,
        disclosedFields: disclosureReq.allowedFields,
        merkleProof: [],
        disclosedRoot: "0x" + "c".repeat(64),
        completeRoot: "0x" + "d".repeat(64),
        verified: true,
        timestamp: Date.now(),
      };

      res.json(proof);
    } catch (error) {
      next(error);
    }
  });

  /**
   * GET /regulator/tamper-evidence/:eventIndex
   * Verify tamper-evidence chain for an event
   */
  router.get("/regulator/tamper-evidence/:eventIndex", async (req: Request, res: Response, next: NextFunction) => {
    try {
      const context = req.regulatorContext as RegulatorContext;
      if (!context) {
        return res.status(401).json({ error: "Unauthorized" });
      }

      const eventIndex = Number(req.params.eventIndex);

      const verification = {
        eventIndex,
        chainValid: true,
        previousEventHash: "0x" + "b".repeat(64),
        currentEventHash: "0x" + "a".repeat(64),
        nextEventHash: "0x" + "e".repeat(64),
        hashAlgorithm: "SHA256",
        timestamp: Date.now(),
        verificationDetails: {
          genesisEvent: false,
          isLastEvent: false,
          chainIntegrityScore: 1.0,
          intermediateEventCount: 100,
        },
      };

      res.json(verification);
    } catch (error) {
      next(error);
    }
  });

  /**
   * POST /regulator/data-sharing-agreements
   * Create a new DSA
   */
  router.post("/regulator/data-sharing-agreements", async (req: Request, res: Response, next: NextFunction) => {
    try {
      const context = req.regulatorContext as RegulatorContext;
      if (context.role !== "admin" && context.role !== "officer") {
        return res.status(403).json({ error: "Insufficient permissions" });
      }

      const dsaReq = req.body as DSARequest;

      const dsa = {
        id: uuidv4(),
        dataProvider: dsaReq.dataProvider,
        regulatorAddress: dsaReq.regulatorAddress,
        standards: dsaReq.standards,
        allowedEventTypes: dsaReq.allowedEventTypes,
        role: dsaReq.role,
        effectiveLedger: dsaReq.effectiveLedger,
        expiryLedger: dsaReq.expiryLedger || 0,
        active: true,
        status: "draft",
        createdAt: Date.now(),
        createdBy: context.regulatorId,
        signatureProvider: null,
        signatureRegulator: null,
      };

      res.status(201).json(dsa);
    } catch (error) {
      next(error);
    }
  });

  /**
   * GET /regulator/data-sharing-agreements
   * List DSAs for current regulator
   */
  router.get("/regulator/data-sharing-agreements", async (req: Request, res: Response, next: NextFunction) => {
    try {
      const context = req.regulatorContext as RegulatorContext;

      const dsas = [
        {
          id: uuidv4(),
          dataProvider: "GBBBBBBB...",
          regulatorAddress: "GAAAAAA...",
          standards: ["ISA3000", "SOC2"],
          allowedEventTypes: ["access_control", "authentication"],
          role: "auditor",
          status: "executed",
          active: true,
          effectiveLedger: 1000,
          expiryLedger: 0,
          createdAt: Date.now(),
        },
      ];

      res.json(dsas);
    } catch (error) {
      next(error);
    }
  });

  /**
   * GET /regulator/data-sharing-agreements/:dsaId
   * Get details of a specific DSA
   */
  router.get("/regulator/data-sharing-agreements/:dsaId", async (req: Request, res: Response, next: NextFunction) => {
    try {
      const context = req.regulatorContext as RegulatorContext;
      const dsaId = req.params.dsaId;

      const dsa = {
        id: dsaId,
        dataProvider: "GBBBBBBB...",
        regulatorAddress: "GAAAAAA...",
        standards: ["ISA3000"],
        allowedEventTypes: ["access_control"],
        role: "officer",
        minSensitivity: "internal",
        status: "executed",
        active: true,
        effectiveLedger: 1000,
        expiryLedger: 52560, // 1 year
        createdAt: Date.now(),
        executedAt: Date.now() + 86400000,
      };

      res.json(dsa);
    } catch (error) {
      next(error);
    }
  });

  /**
   * POST /regulator/compliance-reports
   * Generate a compliance report
   */
  router.post("/regulator/compliance-reports", async (req: Request, res: Response, next: NextFunction) => {
    try {
      const context = req.regulatorContext as RegulatorContext;

      const reportReq = req.body as ComplianceReportRequest;

      const report = {
        id: uuidv4(),
        standard: reportReq.standard,
        auditSubject: reportReq.auditSubject,
        issuer: context.regulatorId,
        generatedAt: Date.now(),
        status: "draft",
        eventsExamined: reportReq.eventsExamined,
        objectivesTested: reportReq.objectivesTested,
        controlsOperating: reportReq.controlsOperating,
        controlsDeficient: reportReq.controlsDeficient,
        findingsSummaryHash: "0x" + "f".repeat(64),
      };

      res.status(201).json(report);
    } catch (error) {
      next(error);
    }
  });

  /**
   * GET /regulator/compliance-reports
   * List compliance reports
   */
  router.get("/regulator/compliance-reports", async (req: Request, res: Response, next: NextFunction) => {
    try {
      const context = req.regulatorContext as RegulatorContext;

      const standard = (req.query.standard as string) || "ISA3000";
      const status = (req.query.status as string) || "published";

      const reports = [
        {
          id: uuidv4(),
          standard,
          auditSubject: "COMPANY_NAME",
          issuer: context.regulatorId,
          generatedAt: Date.now(),
          status,
          eventsExamined: 500,
          controlsOperating: 45,
          controlsDeficient: 3,
        },
      ];

      res.json(reports);
    } catch (error) {
      next(error);
    }
  });

  /**
   * GET /regulator/compliance-reports/:reportId
   * Get a specific compliance report
   */
  router.get("/regulator/compliance-reports/:reportId", async (req: Request, res: Response, next: NextFunction) => {
    try {
      const context = req.regulatorContext as RegulatorContext;
      const reportId = req.params.reportId;

      const report = {
        id: reportId,
        standard: "ISA3000",
        auditSubject: "COMPANY_NAME",
        issuer: context.regulatorId,
        generatedAt: Date.now(),
        status: "published",
        publishedAt: Date.now() - 86400000,
        eventsExamined: 500,
        objectivesTested: ["CC6.1", "CC7.1", "A1.1"],
        controlsOperating: 45,
        controlsDeficient: 3,
        findings: [
          {
            controlObjective: "CC6.1",
            finding: "Segregation of duties control is operating effectively",
            evidence: ["0x" + "a".repeat(64)],
          },
        ],
      };

      res.json(report);
    } catch (error) {
      next(error);
    }
  });

  /**
   * POST /regulator/export
   * Export audit data for compliance analysis
   */
  router.post("/regulator/export", async (req: Request, res: Response, next: NextFunction) => {
    try {
      const context = req.regulatorContext as RegulatorContext;

      const exportReq = req.body;
      const format = exportReq.format || "csv"; // csv, json, pdf

      const filename = `audit_export_${Date.now()}.${format}`;

      res.setHeader("Content-Type", format === "json" ? "application/json" : "text/csv");
      res.setHeader("Content-Disposition", `attachment; filename="${filename}"`);

      // In production, stream the export data
      const sampleData = {
        exportedAt: Date.now(),
        format,
        rowCount: 0,
        checksum: "0x" + "0".repeat(64),
      };

      res.json(sampleData);
    } catch (error) {
      next(error);
    }
  });

  /**
   * GET /regulator/access-requests
   * List access requests for current regulator
   */
  router.get("/regulator/access-requests", async (req: Request, res: Response, next: NextFunction) => {
    try {
      const context = req.regulatorContext as RegulatorContext;

      const requests = [
        {
          id: uuidv4(),
          requester: context.regulatorId,
          dataOwner: "COMPANY_ADDRESS",
          standard: "ISA3000",
          eventTypes: ["access_control"],
          status: "pending",
          createdAt: Date.now(),
        },
      ];

      res.json(requests);
    } catch (error) {
      next(error);
    }
  });

  /**
   * POST /regulator/statistics
   * Get audit trail statistics
   */
  router.post("/regulator/statistics", async (req: Request, res: Response, next: NextFunction) => {
    try {
      const context = req.regulatorContext as RegulatorContext;

      const stats = {
        totalEventsAudited: 5000,
        eventsByType: {
          access_control: 1200,
          authentication: 800,
          data_modification: 1500,
          configuration_change: 400,
          system_exception: 100,
        },
        eventsByStandard: {
          ISA3000: 3000,
          SOC2: 2000,
        },
        complianceScore: 0.94,
        flaggedEvents: 25,
        daysAudited: 365,
      };

      res.json(stats);
    } catch (error) {
      next(error);
    }
  });

  return router;
}

/**
 * Regulator authentication middleware
 */
export function regulatorAuthMiddleware(req: Request, res: Response, next: NextFunction) {
  try {
    const bearerToken = req.headers.authorization?.split(" ")[1];
    if (!bearerToken) {
      return res.status(401).json({ error: "Missing authorization header" });
    }

    // In production, verify JWT and extract regulator context
    req.regulatorContext = {
      regulatorId: "REG_001",
      role: "auditor",
      standards: ["ISA3000", "SOC2"],
    };

    next();
  } catch (error) {
    res.status(401).json({ error: "Invalid authentication" });
  }
}

File: src/regulator.rs
//! Regulator-specific audit trail extensions with selective disclosure,
//! data sharing agreements, and compliance standards (ISA 3000, SOC2).
//!
//! This module provides:
//! - Regulator-specific event types and compliance classifications
//! - Tamper-evidence verification via hash chains
//! - Selective disclosure with cryptographic proofs
//! - Data sharing agreement management
//! - Compliance standard validators (ISA 3000, SOC2)

use soroban_sdk::{contracttype, BytesN, Address, Symbol, Bytes, Vec, Env};

/// Compliance audit standards supported by the system
#[contracttype]
#[derive(Clone, Copy, Debug, Eq, PartialEq, PartialOrd, Ord)]
#[repr(u32)]
pub enum ComplianceStandard {
    /// ISA 3000: International Standard on Assurance Engagements
    ISA3000 = 0,
    /// SOC 2: Service Organization Control Framework
    SOC2 = 1,
    /// GDPR compliance framework
    GDPR = 2,
    /// SOX: Sarbanes-Oxley Act
    SOX = 3,
}

/// Regulator role and permission levels
#[contracttype]
#[derive(Clone, Copy, Debug, Eq, PartialEq, PartialOrd, Ord)]
#[repr(u32)]
pub enum RegulatorRole {
    /// View-only access to audit events
    Auditor = 0,
    /// Can request disclosures and manage access
    RegulatorOfficer = 1,
    /// Full administrative access
    RegulatoryAdmin = 2,
}

/// Data sensitivity level for selective disclosure
#[contracttype]
#[derive(Clone, Copy, Debug, Eq, PartialEq, PartialOrd, Ord)]
#[repr(u32)]
pub enum SensitivityLevel {
    /// Public information
    Public = 0,
    /// Internal use only
    Internal = 1,
    /// Confidential
    Confidential = 2,
    /// Highly restricted
    Restricted = 3,
}

/// Event classification for regulatory purposes
#[contracttype]
#[derive(Clone, Debug, Eq, PartialEq)]
pub struct RegulatoryEventClass {
    /// Standard this classification belongs to (ISA 3000, SOC2, etc.)
    pub standard: ComplianceStandard,
    /// Classification code (e.g., "CC6.1", "A1.1" for ISA 3000)
    pub control_code: Symbol,
    /// Whether this event demonstrates control effectiveness
    pub demonstrates_control: bool,
    /// Required retention period in ledgers
    pub retention_ledgers: u32,
    /// Sensitivity level for selective disclosure
    pub sensitivity: SensitivityLevel,
}

/// Tamper-evidence proof for a single event
#[contracttype]
#[derive(Clone, Debug, Eq, PartialEq)]
pub struct TamperProof {
    /// Event index in the global log
    pub event_index: u32,
    /// SHA-256 hash of this event
    pub event_hash: BytesN<32>,
    /// SHA-256 hash of the previous event
    pub prev_hash: BytesN<32>,
    /// Timestamp of this event
    pub timestamp: u64,
    /// Timestamp of the next event (0 if not found)
    pub next_timestamp: u64,
    /// Whether the hash chain is valid (event_hash correctly computed)
    pub chain_valid: bool,
}

/// Selective disclosure proof for an event
/// Proves certain facts about an event without revealing full metadata
#[contracttype]
#[derive(Clone, Debug, Eq, PartialEq)]
pub struct SelectiveDisclosureProof {
    /// The event being disclosed
    pub event_index: u32,
    /// Merkle root of disclosed fields
    pub disclosed_root: BytesN<32>,
    /// Merkle root of all event fields
    pub complete_root: BytesN<32>,
    /// List of disclosed field names (e.g., ["timestamp", "submitter"])
    pub disclosed_fields: Vec<Symbol>,
    /// Cryptographic proof of inclusion for disclosed fields
    pub merkle_proof: Vec<BytesN<32>>,
}

/// Data Sharing Agreement between entity and regulator
#[contracttype]
#[derive(Clone, Debug, Eq, PartialEq)]
pub struct DataSharingAgreement {
    /// Unique identifier for this agreement
    pub id: BytesN<32>,
    /// Entity providing the audit data
    pub data_provider: Address,
    /// Regulator receiving access
    pub regulator_address: Address,
    /// When this agreement takes effect (ledger sequence number)
    pub effective_ledger: u32,
    /// When this agreement expires (0 = no expiry)
    pub expiry_ledger: u32,
    /// Allowed compliance standards under this agreement
    pub standards: Vec<ComplianceStandard>,
    /// Permissible event types to access
    pub allowed_event_types: Vec<Symbol>,
    /// Regulator role granted
    pub role: RegulatorRole,
    /// Minimum sensitivity level that can be disclosed
    pub min_sensitivity: SensitivityLevel,
    /// Whether agreement is currently active
    pub active: bool,
    /// Signature of the agreement by both parties
    pub signature_provider: BytesN<64>,
    pub signature_regulator: BytesN<64>,
}

/// Regulator portal access request
#[contracttype]
#[derive(Clone, Debug, Eq, PartialEq)]
pub struct AccessRequest {
    /// Unique request ID
    pub id: BytesN<32>,
    /// Regulator requesting access
    pub requester: Address,
    /// Entity whose data is requested
    pub data_owner: Address,
    /// Compliance standard for which access is needed
    pub standard: ComplianceStandard,
    /// Requested event types to access
    pub event_types: Vec<Symbol>,
    /// Legal basis for the request (e.g., regulatory authority, audit mandate)
    pub legal_basis: Bytes,
    /// Proposed data sharing agreement terms
    pub proposed_terms: DataSharingAgreement,
    /// Request status: 0=pending, 1=approved, 2=rejected
    pub status: u32,
    /// Timestamp when request was created
    pub created_at: u64,
    /// Timestamp when request was resolved (0 if pending)
    pub resolved_at: u64,
}

/// Compliance audit report generated for a regulator
#[contracttype]
#[derive(Clone, Debug, Eq, PartialEq)]
pub struct ComplianceReport {
    /// Report identifier
    pub id: BytesN<32>,
    /// Report type (ISA 3000, SOC2, etc.)
    pub standard: ComplianceStandard,
    /// Entity being audited
    pub audit_subject: Address,
    /// Regulator issuing report
    pub issuer: Address,
    /// Report generation date (ledger timestamp)
    pub generated_at: u64,
    /// Report status: 0=draft, 1=pending_review, 2=published, 3=withdrawn
    pub status: u32,
    /// Total events examined
    pub events_examined: u32,
    /// Control objectives tested
    pub objectives_tested: Vec<Symbol>,
    /// Controls found to be operating effectively
    pub controls_operating: u32,
    /// Controls found deficient
    pub controls_deficient: u32,
    /// Key findings (as opaque bytes for privacy)
    pub findings_summary_hash: BytesN<32>,
}

/// Audit trail query filter for regulators
#[contracttype]
#[derive(Clone, Debug, Eq, PartialEq)]
pub struct AuditFilter {
    /// Event types to include (empty = all)
    pub event_types: Vec<Symbol>,
    /// Start timestamp (inclusive)
    pub start_time: u64,
    /// End timestamp (inclusive)
    pub end_time: u64,
    /// Submitter address filter (empty = all)
    pub submitter: Option<Address>,
    /// Minimum sensitivity to include
    pub min_sensitivity: SensitivityLevel,
    /// Whether to include only control-demonstrating events
    pub only_control_events: bool,
}

/// Audit log entry with regulator metadata
#[contracttype]
#[derive(Clone, Debug, Eq, PartialEq)]
pub struct RegulatorAuditEntry {
    /// Event index in global log
    pub event_index: u32,
    /// Event hash for verification
    pub event_hash: BytesN<32>,
    /// Event timestamp
    pub timestamp: u64,
    /// Event type
    pub event_type: Symbol,
    /// Submitter address
    pub submitter: Address,
    /// Regulatory classification
    pub regulatory_class: Option<RegulatoryEventClass>,
    /// Sensitivity level
    pub sensitivity: SensitivityLevel,
    /// Whether this event demonstrates control effectiveness
    pub control_event: bool,
}

/// Enum for regulator contract errors
#[derive(Clone, Copy, Debug, Eq, PartialEq, PartialOrd, Ord)]
#[repr(u32)]
pub enum RegulatorError {
    /// Regulator not authorized for this operation
    UnauthorizedRegulator = 100,
    /// Data Sharing Agreement not found
    AgreementNotFound = 101,
    /// Agreement has expired
    AgreementExpired = 102,
    /// Insufficient permissions for requested access
    InsufficientPermissions = 103,
    /// Event type not covered by agreement
    EventTypeNotAllowed = 104,
    /// Selective disclosure proof invalid
    InvalidDisclosureProof = 105,
    /// Tamper-evidence chain broken
    TamperEvidenceViolation = 106,
    /// Compliance standard not supported
    UnsupportedStandard = 107,
    /// Access request not found
    AccessRequestNotFound = 108,
    /// Invalid signature on agreement
    InvalidAgreementSignature = 109,
    /// Regulator role not recognized
    InvalidRole = 110,
}

/// Helper struct for ISA 3000 compliance validation
#[contracttype]
#[derive(Clone, Debug, Eq, PartialEq)]
pub struct ISA3000ControlObjective {
    /// Control objective code (e.g., "CC6.1", "A1.1")
    pub code: Symbol,
    /// Description of the objective
    pub description: Bytes,
    /// Related audit evidence event types
    pub evidence_types: Vec<Symbol>,
    /// Whether objective requires continuous monitoring
    pub continuous_monitoring: bool,
}

/// Helper struct for SOC 2 compliance validation
#[contracttype]
#[derive(Clone, Debug, Eq, PartialEq)]
pub struct SOC2Criterion {
    /// Criterion code (e.g., "CC6.1", "A1.1")
    pub code: Symbol,
    /// Trust service principle (Security, Availability, Processing Integrity, Confidentiality, Privacy)
    pub principle: Bytes,
    /// Description of the criterion
    pub description: Bytes,
    /// Related audit evidence event types
    pub evidence_types: Vec<Symbol>,
}

/// Container for compliance rule sets
#[contracttype]
#[derive(Clone, Debug, Eq, PartialEq)]
pub struct ComplianceRuleSet {
    /// Standard this ruleset applies to
    pub standard: ComplianceStandard,
    /// Version of the ruleset
    pub version: u32,
    /// Whether this ruleset is active
    pub active: bool,
    /// Rules encoded as opaque bytes (specific format per standard)
    pub rules_data: Bytes,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_compliance_standard_ordering() {
        assert!(ComplianceStandard::ISA3000 < ComplianceStandard::SOC2);
        assert!(ComplianceStandard::SOC2 < ComplianceStandard::GDPR);
    }

    #[test]
    fn test_regulator_role_ordering() {
        assert!(RegulatorRole::Auditor < RegulatorRole::RegulatorOfficer);
        assert!(RegulatorRole::RegulatorOfficer < RegulatorRole::RegulatoryAdmin);
    }

    #[test]
    fn test_sensitivity_level_ordering() {
        assert!(SensitivityLevel::Public < SensitivityLevel::Internal);
        assert!(SensitivityLevel::Internal < SensitivityLevel::Confidential);
        assert!(SensitivityLevel::Confidential < SensitivityLevel::Restricted);
    }
}
